### PR TITLE
Docker: docker images, microservices.

### PR DESCRIPTION
**Pull request by Durf1985**
## In the process done:

* Assemble Docker images for a service application
* Optimize work with Docker images
* Break our application into several components
* Launch and operation of an application based on Docker images
* Launch our microservice application

## How to setup the project:

* Switch to GCP VM environment

```
eval $(docker-machine env name_your_docker_host)
```

* Build images 

```
cd your_repo/src
docker build -t <your-dockerhub-login>/ui:1.0 ./ui
docker build -t <your-dockerhub-login>/comment:1.0 ./comment
docker build -t <your-dockerhub-login>/post:1.0 ./post-py
```

* Create bridge network

```
docker network create your_name
```

* Run containers

```
docker run -d --network=reddit --network-alias=post_db --network-alias=comment_db mongo:3.2
docker run -d --network=reddit --network-alias=post <your-dockerhub-login>/post:1.0
docker run -d --network=reddit --network-alias=comment <your-dockerhub-login>/comment:1.0
docker run -d --network=reddit -p 9292:9292 <your-dockerhub-login>/ui:1.0
```

## How to check the project:

* Use the docker-machine ls command to find out the ip address of your virtual machine
* Follow the link <http://your_vm_ip_address:9292>

## PR checklist

The label with the homework number, in label description - homework lecture and topic

### CODE REVIEW

File: src/comment/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM ubuntu:18.04
+RUN apt-get update -qq \
+&& apt-get install -y --no-install-recommends \
```
Comment (loktionovam) [inline src/comment/Dockerfile]: use indentation here

File: src/ui/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+FROM ubuntu:18.04
+RUN apt-get update -qq \
+&& apt-get install -y --no-install-recommends \
```
Comment (loktionovam) [inline src/ui/Dockerfile]: use indentation

File: src/post-py/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,12 @@
+FROM python:3.6.0-alpine
+
+WORKDIR /app
+COPY . /app
```
Comment (loktionovam) [inline src/post-py/Dockerfile]: This code is changed frequently and it will reset your docker images cache so move it below installation of dependencies

**Review by loktionovam — CHANGES_REQUESTED** [review]

File: src/post-py/Dockerfile (hunk 3 above)
Comment (Durf1985) [reply src/post-py/Dockerfile]: In this case, this option copies the file `requirements.txt` inside the image, pip install will not work without this file.

Comment (loktionovam) [reply src/post-py/Dockerfile]: copy requirements.txt separately and install the dependencies first (they are changed not so frequently as application code)

File: src/ui/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+RUN bundle install
+COPY . $APP_HOME
+
+ENV POST_SERVICE_HOST post
+ENV POST_SERVICE_PORT 5000
+ENV COMMENT_SERVICE_HOST comment
+ENV COMMENT_SERVICE_PORT 9292
+
+FROM alpine:3.5
+ENV USER Durf
```
Comment (loktionovam) [inline src/ui/Dockerfile]: Parametrisation is excessive here and can lead to errors while using image

File: src/ui/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+RUN apk --no-cache add \
+    ruby=2.3.8-r0 \
+    ruby-bundler=1.13.4-r0 \
+    ruby-json=2.3.8-r0
+
+COPY --from=builder $GEM_HOME $GEM_HOME
+COPY --from=builder $APP_HOME $APP_HOME
+WORKDIR $APP_HOME
+
+RUN adduser -D $USER
```
Comment (loktionovam) [inline src/ui/Dockerfile]: You can add this command above (RUN apk ...) and remove unnecessary layer

**Review by loktionovam — CHANGES_REQUESTED** [review]

**Review by loktionovam — APPROVED** [review]

**Pull request merged**